### DISCUSSION TITLE
Issue with strings.h and application that defines its own "mcpy" function.

### DISCUSSION THREAD

**Discussion by fswarbrick** [zopencommunity/meta#845] [Porting]
So I am porting JOE (Joe's Own Editor) and running in to an annoyance.  JOE defines a function named mcpy.  When a file that includes strings.h and has its own mcpy() declaration is compiled it gets an error that there are `conflicting types for 'mcpy'`.  This is because strings.h contains a declaration for a function called mcpy. 

If the file that includes strings.h is compiled with no special macros defined at compile time this error does not occur.  However, it looks like ZOT includes both  `-D_XOPEN_SOURCE=600` and `-D_ALL_SOURCE`, which in turn sets the __UU macro.  If the __UU macro is defined then the declaration of mcpy is done when strings.h is included.

The first question I have is what is the point of strings.h declaring mcpy.  It does not appear to be a member in the z/OS runtime libraries.  I wonder if it is a legacy declaration in strings.h that is no longer required.  I realize this would be a z/OS base issue, not a ZOT issue.

In any case, any thoughts on the best way to get around this?  The way I've done it at this point is to copy /usr/include/strings.h in to the application's source directory, and then remove the declaration.  With this and a few other minor changes I have gotten JOE to function.

If you want to try it yourself, here is a tiny test program that demonstrates the issue.
```
#include <stddef.h>
#include <strings.h>

void *mcpy(void *a, const void *b, ptrdiff_t len);

int main() {
        char s[] = "test";
        char s2[80];

        mcpy(s2, s, sizeof(s));
}
```
And compile with
```clang -D_XOPEN_SOURCE=600 -D_ALL_SOURCE memcp.c``` (obviously replacing clang with whatever C compiler you are using)

You will get results looking something like this:
```
memcp.c:4:7: error: conflicting types for 'mcpy'
void *mcpy(void *a, const void *b, ptrdiff_t len);
      ^
/usr/include/strings.h:203:17: note: previous declaration is here
         void * mcpy (void *, const void *, size_t);
                ^
1 error generated.
```
Without specifying the two macro definitions it compiles fine (but the link-edit fails because mcpy does not exist).

**Comment by IgorTodorovskiIBM** [comment]
Interesting, it's guarded out for xlclang, but not clang:
```
202        #elif !defined(__ibmxl__)
203          void * mcpy (void *, const void *, size_t);
```
FYI @perry-ca 

One solution is to use a trick like this: https://github.com/ibmruntimes/zoslib/blob/zopen2/include/stdio.h#L33-L36

**Reply by perry-ca** [reply]
The guard is correct.  The block above is for XL.  The elif is for OpenXL.  The problem is the first two chars are missing.  The line should be:
```
void * memcpy (void *, const void *, size_t);
```
We'll need to ask LE to fix this.

**Reply by fswarbrick** [reply]
Haha, well that makes more sense.  I should have noticed that.
Will you make the request to LE to fix?

**Reply by fswarbrick** [reply]
Simple enough.
Created strings.h in the projects source folder:
```
#undef mcpy
#define mcpy __mcpy_dummy

#include_next <strings.h>
#undef mcpy
```
Thanks!

**Reply by perry-ca** [reply]
I'd define mcpy to memcpy so you get the correct declaration.

**Reply by fswarbrick** [reply]
The application doesn't appear to use it, so I didn't bother.

**Reply by IgorTodorovskiIBM** [reply]
And let me know if you want to contribute it as part of z/OS Open Tools, we'd be happy to set up a repo for JOE.

**Reply by fswarbrick** [reply]
@IgorTodorovskiIBM   Yes, I'd like that.

**Reply by IgorTodorovskiIBM** [reply]
Great! https://github.com/ZOSOpenTools/joeport - you should have an invitation for access

**Reply by IgorTodorovskiIBM** [reply]
Are you on the System Z enthusiasts discord channel?

**Reply by fswarbrick** [reply]
I am not on that discord.

**Reply by IgorTodorovskiIBM** [reply]
If you're interested in joining, here's the invite: https://discord.gg/FhjY4tNH .
